tuf/api/metadata.py

```diff
@@ -116,6 +116,33 @@ def __init__(self, signed: T, signatures: Dict[str, Signature]):
         self.signed: T = signed
         self.signatures = signatures
 
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Metadata):
+            return False
+
+        # Verifies that the signatures for both objects are the same.
+        # TODO: Use Signature.__eq__ (secure-systems-lab/securesystemslib#383)
+        # pylint: disable=unidiomatic-typecheck
+        if not (type(self.signatures) == type(other.signatures) == dict):
+            return False
+
+        # Verifies that signatures keyids for both objects are the same
+        if self.signatures.keys() != other.signatures.keys():
+            return False
+
+        # Iterate over self.signatures and other.signatures at the same time
+        # as the signatures in the file format are ordered.
+        keyids = zip(self.signatures.keys(), other.signatures.keys())
```

Comment:
After the check `if self.signatures.keys() != other.signatures.keys():` we know `self` and `other` have the same keys, but we don't know if they follow the same order.

When I zip them I make sure I traverse both of them following their corresponding insertion order.
If the `self.signatures` and `other.signatures` orders are different, then `self_keyid != other_keyid` will return False.

I tried without zipping, but it fails the test `test_md_eq_signatures_reversed_order` inside `test_metadata_eq_.py`.

Reply:
> After the check if self.signatures.keys() != other.signatures.keys(): we know that self and other have the same keys, but we don't know if they follow the same order.

but doesn't that check actually tell us that the order is correct: List eq does check the order?

Then that would mean you don't need the loop at all -- could just check `self.signatures != other.signatures` since now you know the order is the same

Reply:
Ah right, good point, because `keys()` doesn't return a list anymore in Python3 but rather something set-like and thus unordered.

But still, you're comparing keyids twice here. You could instead only check the lengths at first, i.e.  `len(self.signatures.keys()) != len(other.signatures.keys())`, or just pass `strict=True` to `zip`, which does the same thing.

Reply:
>  could just check self.signatures != other.signatures

Not until https://github.com/secure-systems-lab/securesystemslib/pull/383 is merged and release.

Reply:
> > could just check self.signatures != other.signatures
> 
> Not until [secure-systems-lab/securesystemslib#383](https://github.com/secure-systems-lab/securesystemslib/pull/383) is merged and released.

... maybe we should just go ahead and do that, before spending more thoughts on `zip` :D